set/validate object namespace before admission

liggitt opened the pull request:
**What type of PR is this?**

/kind bug
/kind cleanup

**What this PR does / why we need it**:
Ensures the namespace population/check that occurs before an object is persisted is done before it is sent to admission. This ensures the namespace in the object matches the request namespace attribute.

Currently, namespaced objects can be sent to admission with empty namespaces during creation (and are defaulted to match the request namespace just before persisting) or a mismatched namespace (and are rejected just before persisting).

Added tests around the following scenarios:
* cluster-scoped, no namespace in object
* cluster-scoped, mismatched namespace in object
* namespaced, no namespace in object
* namespaced, matching namespace in object
* namespaced, mismatched namespace in object

xref https://github.com/kubernetes/kubernetes/issues/88282

**Does this PR introduce a user-facing change?**:
```release-note
kube-apiserver: ensures the namespace of objects sent to admission webhooks matches the request namespace. Previously, objects without a namespace set would have the request namespace populated after mutating admission, and objects with a namespace that did not match the request namespace would be rejected after admission.
```

/cc @deads2k

liggitt committed:
set/validate object namespace before admission

lavalamp commented:
/assign @deads2k 
cc @MikeSpreitzer @yue9944882

MikeSpreitzer commented:
Do not forget that CRDs are still not fully up to the job of obviating the original way of defining new resources using aggregated custom apiservers.  Some of us are doing that, with a distinct etcd cluster for the custom resources too.  Now think about backup/restore of the etcd clusters.  It would be best if there is a rigorous independence of their content.  It might be best to not require one etcd cluster to contain a namespace object in order to persist an object in a different etcd cluster.

liggitt commented:
> Do not forget that CRDs are still not fully up to the job of obviating the original way of defining new resources using aggregated custom apiservers. Some of us are doing that, with a distinct etcd cluster for the custom resources too. Now think about backup/restore of the etcd clusters. It would be best if there is a rigorous independence of their content. It might be best to not require one etcd cluster to contain a namespace object in order to persist an object in a different etcd cluster.

That is unrelated to this PR.

The API server currently ensures the namespace in the request path and the namespace in the object match prior to persisting in etcd, but that check is done post-admission. This PR ensures that check is done prior to admission to avoid sending namespaced objects to admission with their namespace field unset (to be filled in later prior to persistence) or mismatching the request (to be rejected later prior to persistence).

caesarxuchao commented:
/triage accepted

liggitt commented:
/hold for tests

sftim commented on staging/src/k8s.io/apiserver/pkg/registry/rest/meta_test.go:
```diff
@@ -54,31 +54,3 @@ func TestHasObjectMetaSystemFieldValues(t *testing.T) {
 	}
 }
-
-// TestValidNamespace validates that namespace rules are enforced on a resource prior to create or update
-func TestValidNamespace(t *testing.T) {
-	ctx := genericapirequest.NewDefaultContext()
-	namespace, _ := genericapirequest.NamespaceFrom(ctx)
-	// TODO: use some genericapiserver type here instead of clientapiv1
-	resource := example.Pod{}
-	if !ValidNamespace(ctx, &resource.ObjectMeta) {
-		t.Fatalf("expected success")
-	}
-	if namespace != resource.Namespace {
-		t.Fatalf("expected resource to have the default namespace assigned during validation")
-	}
-	resource = example.Pod{ObjectMeta: metav1.ObjectMeta{Namespace: "other"}}
-	if ValidNamespace(ctx, &resource.ObjectMeta) {
-		t.Fatalf("Expected error that resource and context errors do not match because resource has different namespace")
-	}
-	ctx = genericapirequest.NewContext()
-	if ValidNamespace(ctx, &resource.ObjectMeta) {
-		t.Fatalf("Expected error that resource and context errors do not match since context has no namespace")
-	}
-
-	ctx = genericapirequest.NewContext()
-	ns := genericapirequest.NamespaceValue(ctx)
-	if ns != "" {
-		t.Fatalf("Expected the empty string")
-	}
-}
```
Do there need to be any new tests to stand in place of these ones?

liggitt replied:
yup, hence https://github.com/kubernetes/kubernetes/pull/94637#issuecomment-974701175 :)

sftim replied:
OK, wasn't sure if you meant “This isn't complete until I write new tests” or “I want to see if the existing tests fail”.

liggitt commented:
updated with tests

liggitt commented:
/hold cancel

dims commented:
/retest

dims commented:
/assign

liggitt commented:
the unit test failures are legitimate... there's a lot of unit tests not populating the context with the same namespace info a real API request would have gotten :-/

dims commented:
Ack @liggitt

MadhavJivrajani commented:
@liggitt 👋
I should be able to help out with updating the unit tests next week :)

liggitt commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go:
```diff
@@ -152,7 +152,7 @@ func createHandler(r rest.NamedCreater, scope *RequestScope, admit admission.Int
 			_, name, _ = scope.Namer.ObjectName(obj)
 		}
-		if len(namespace) == 0 && *gvk == namespaceGVK {
+		if len(namespace) == 0 && scope.Resource == namespaceGVR {
```
this is existing, but comparing to the request scope resource rather than the decoded object Kind is more correct

MadhavJivrajani committed:
Update tests to have namespace set in req context

* This updates tests that make requests that call
  into admission.
* For apiextensions-apiserver, the namespace is set
  to metav1.NamespaceNone since the CRD created and
  worked with is a cluster scoped CRD.
* For most tests in register/storage, the namespace
  is set to NamespaceDefault or NamespaceNone based
  on if the resource is cluster scoped (namespaces)
  or namespace scoped (pvs).
* Endpoints tests now have a namespace specified in
  the expected object - NamespaceDefault or other
  based on the test.

Signed-off-by: Madhav Jivrajani <madhav.jiv@gmail.com>

liggitt commented:
pulled in unit test fixes from https://github.com/kubernetes/kubernetes/pull/108068

MadhavJivrajani commented on pkg/registry/core/persistentvolume/storage/storage_test.go:
```diff
@@ -169,7 +169,7 @@ func TestUpdateStatus(t *testing.T) {
 	defer server.Terminate(t)
 	defer storage.Store.DestroyFunc()
-	ctx := genericapirequest.NewContext()
+	ctx := genericapirequest.NewDefaultContext()
```
```suggestion
	ctx := genericapirequest.WithNamespace(genericapirequest.NewContext(), metav1.NamespaceNone)
```
From https://github.com/kubernetes/kubernetes/pull/108068#discussion_r804682973
(I hadn't pushed suggested changes on account of https://github.com/kubernetes/kubernetes/pull/108068#discussion_r808130936

liggitt replied:
ah, right... I squashed the comments into your commit, and split out the test bit that *has* to be part of this PR because it relies on the create/update handler changes

I think the current `Update tests to have namespace set in req context` commit in this PR could actually pass tests and merge to master ahead of this PR if you want to pull it back out... sorry for the churn :)

MadhavJivrajani replied:
Nope, this looks good! Thanks!
The only thing I can think of is my commit message which contains details pre-your-review, you might have to one last rebase to change that 😬

liggitt committed:
Update storage/handler unit tests to set / expect namespace in object

liggitt commented:
/hold for api-machinery review as well

fedebongio commented:
/assign @jpbetz 
Joe would you be able to take a look? 
/remove-triage accepted

jpbetz commented:
/triage accepted

deads2k commented on pkg/registry/core/namespace/storage/storage_test.go:
```diff
@@ -167,6 +167,7 @@ func TestDeleteNamespaceWithIncompleteFinalizers(t *testing.T) {
 	if err := storage.store.Storage.Create(ctx, key, namespace, nil, 0, false); err != nil {
 		t.Fatalf("unexpected error: %v", err)
 	}
+	ctx = genericapirequest.WithNamespace(genericapirequest.NewContext(), namespace.Name)
```
note to self to look up the current state of namespace value on the context of request for namespaces.

liggitt replied:
see `request.WithNamespace` calls inside endpoints/handlers/* for current state

deads2k commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go:
```diff
@@ -163,6 +163,17 @@ func createHandler(r rest.NamedCreater, scope *RequestScope, admit admission.Int
 		userInfo, _ := request.UserFrom(ctx)
 
+		// ensure namespace on the object is correct, or error if a conflicting namespace was set in the object
+		objectMeta, err := meta.Accessor(obj)
```
breadcrumbs for me to chase.  I don't recall us previously enforcing objects to have objectmeta, so https://github.com/openshift/api/blob/master/authorization/v1/types.go#L238-L243 doesn't have it for instance. Perhaps it overrides this entire flow?  i wouldn't expect it to because that would bypass audit, but maybe.

liggitt replied:
fair... if we can't get an accessor, I think it's reasonable to skip this here.

We still have the backstop in staging/src/k8s.io/apiserver/pkg/registry/rest/create.go and staging/src/k8s.io/apiserver/pkg/registry/rest/update.go that require Accessor and repeat the `EnsureObjectNamespaceMatchesRequestNamespace` check before persisting regular objects (types without objectmeta that implement their own Create impl skip *that* flow)

deads2k commented on staging/src/k8s.io/apiserver/pkg/registry/rest/meta.go:
```diff
@@ -31,13 +30,44 @@ func FillObjectMetaSystemFields(meta metav1.Object) {
+
+	case requestNamespace == metav1.NamespaceNone:
+		// cluster-scoped, clear namespace
+		obj.SetNamespace(metav1.NamespaceNone)
```
interesting.  We used to clear this?

liggitt replied:
yeah, see BeforeCreate/BeforeUpdate

deads2k commented on staging/src/k8s.io/apiserver/pkg/registry/rest/update.go:
```diff
@@ -110,12 +111,14 @@ func BeforeUpdate(strategy RESTUpdateStrategy, ctx context.Context, obj, old run
-			return errors.NewBadRequest("the namespace of the provided object does not match the namespace sent on the request")
-		}
-	} else if len(objectMeta.GetNamespace()) > 0 {
-		objectMeta.SetNamespace(metav1.NamespaceNone)
```
so we did.  weird.

deads2k commented on staging/src/k8s.io/apiserver/pkg/registry/rest/meta.go:
```diff
@@ -31,13 +30,44 @@ func FillObjectMetaSystemFields(meta metav1.Object) {
-func ValidNamespace(ctx context.Context, resource metav1.Object) bool {
-	ns, ok := genericapirequest.NamespaceFrom(ctx)
-	if len(resource.GetNamespace()) == 0 {
-		resource.SetNamespace(ns)
```
how on earth did we ever write a ValidFoo method that mutates.  I'm too afraid to check blame.

deads2k commented:
/lgtm
/hold cancel
